Add copy on long press to device details in Form Metdata

seadowg opened the pull request:
Closes #3633
Closes #3646

This makes it possible to copy any of the device defined metadata values on the "Form Metadata" screen by long pressinging and then hitting copy. 

There are a couple of code changes that went in as well:
* Added `androidx.test` runner and `androidx.test.fragment` testing utils dependencies. These let us use `FragmentScenario` to write tests for individual `Fragment` objects. Behind the scenes this uses Robolectric.
* Moved `SharedPreferencesTest` from `androidTest` to `test`
* Deprecate `GeneralSharedPreferences.getInstance()` and `AdminSharedPreferences.getInstance()`. Static singletons like this cause problems in tests as the same instance is retained between tests. The singleton is now managed in Dagger (which has a dependency module attached to the lifecycle of the `Application`, not the process).
 
#### What has been done to verify that this works as intended?

#### Why is this the best possible solution? Were any other approaches considered?

Getting this to work meant adding in the new `androix.preference` framework. This is replacing the deprecated `android.preference` framework which didn't include the copy feature. This definitely felt like the right solution vs hacking the feature in ourselves but it does mean we needed to create a new activity (`FormMetadataPreferencesActivity`) that works with this framework. I think we can slowly move to using this for every preference screen - probably just by updating them as we touch them.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Only thing I'd be worried about is regressions on the Form Metadata screen itself so good to have a play with that.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

seadowg committed:
Use androidx preferences for Form Metadata

seadowg committed:
Enable copying on Form Metadata device details

seadowg committed:
Make Form Metadata titles more human

seadowg committed:
Make install ID pref non persistable

seadowg committed:
Rename new prefs activity layout

seadowg committed:
Use summary provider to load preferences managed by PropertyManager

seadowg committed:
Stop rotation creating problems for FormMetadataFragment

seadowg committed:
Stop using deprecated method for Dagger injection

seadowg committed:
Remove margin from preferences

seadowg committed:
Make sure installID lives outside of general prefs so it's not reset

seadowg committed:
Clear shared prefs before running tests

seadowg committed:
Use Dagger to manage singleton of GeneralSharedPreferences

seadowg committed:
Use Dagger to manage singleton of AdminSharedPreferences

lognaturel commented on collect_app/src/main/AndroidManifest.xml:
```diff
@@ -285,5 +285,6 @@ the specific language governing permissions and limitations under the License.
                 <category android:name="android.intent.category.DEFAULT" />
             </intent-filter>
         </activity>
+        <activity android:name=".preferences.FormMetadataPreferencesActivity" />
```
This should be above the `External API` comment. I'd tend to group it with others in the preferences package.

seadowg replied:
Ah sorry yes this was just auto generated. But good catch as that's how these files get so messy!

lognaturel commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -484,8 +484,8 @@
     <string name="email">Email address</string>
-    <string name="user_defined">User-defined</string>
-    <string name="device_defined">Device-defined</string>
+    <string name="user_defined">User defined</string>
```
Does the hyphen seem wrong to you? These are compound words so the hyphen is required at least by [APA guidelines](https://apastyle.apa.org/learn/faqs/when-use-hyphen). Maybe it's a regional thing?

Changing these strings means they'll be pushed out to translators. If translators don't re-translate them before the release then they'll show up in English instead.

seadowg replied:
Hmmm yeah it feels wrong to me as a title (rather than in a sentence). Are we using that as a language guide for English? Would be good to reference that.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/injection/config/AppDependencyModule.java:
```diff
@@ -166,7 +165,7 @@ public ActivityAvailability providesActivityAvailability(Context context) {
     @Provides
     InstallIDProvider providesInstallIDProvider(Context context) {
-        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
+        SharedPreferences preferences = context.getSharedPreferences("meta", Context.MODE_PRIVATE);
```
I think that this is good but I'm concerned that it will get lost and forgotten. How about marking `KEYS_WE_SHOULD_NOT_RESET` as deprecated and describing this strategy? Maybe `meta` should be a constant somewhere?

seadowg replied:
Yeah I was in two minds about this as I'm not a fan of the existing shared prefs abstractions and the keys classes. I do agree that it would be nice to have something that makes the `meta` prefs more discoverable though so they get used for other things. I'll have a think.

seadowg replied:
As a note I'd like to keep the test with a hardcoded/inline as this prevents someone changing the constant (in the code) and the test still passing. It's checking configuration as well as behaviour.

seadowg replied:
I'm going to go with a provider for the `meta` prefs for now. I didn't want to start building a whole `MetaSharedPrefences` to mimic the other ones. I like that we have abstraction for prefs storage though so I might follow up with a refactor that pulls out the shared logic around shared prefs and maybe improves the interfaces a little.

lognaturel reviewed (commented):
Couple of small things that I think should be addressed. Please mark for QA after!

lognaturel reviewed (changes requested):
Should have marked as `request changes` even if it's small things.

seadowg committed:
Move FormMetadataPreferencesActivity declaration

seadowg committed:
Revert "Make Form Metadata titles more human"

This reverts commit 0b0e99b9ae21bbd4bcb03833bd0dacc9fec03e9b.

seadowg committed:
Add MetaSharedPreferencesProvider

mmarciniak90 commented:
@seadowg I noticed unexpected behavior. 

1. Click on `Username` or `Phone number` or `Email address` to open dialog
2. Close the dialog
3. `SIM serial number` or `Subscriber ID` become inactive

|master | first run on PR| after first opening some dialog | after the second opening dialog|
|---|----|----|---|
|![Screenshot_20200221-122544](https://user-images.githubusercontent.com/3907554/75033306-e87e8280-54aa-11ea-93de-247f36c0dbee.png)|![Screenshot_20200221-123134](https://user-images.githubusercontent.com/3907554/75033314-ec120980-54aa-11ea-96aa-4c05697696df.png)|![Screenshot_20200221-123142](https://user-images.githubusercontent.com/3907554/75033323-ef0cfa00-54aa-11ea-8637-2784540163a4.png)|![Screenshot_20200221-123146](https://user-images.githubusercontent.com/3907554/75033328-f16f5400-54aa-11ea-888a-4b400c569ab6.png)|

It is not reproducible in 100% but it is visible very often. It also depends on the device. For me, reproducing this case on Android 7 is the easiest.

Rotating or reopening this page restore default appearance.

Is it possible that these changes cause an impact on changing the language on Android 7?
I cannot change it to e.g. Spanish but I can't reproduce it on the master branch. 

@opendatakit-bot unlabel "needs testing"

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

seadowg commented:
@mmarciniak90 hmmm. Yeah it could be that the new Preferences framework causes problems with language changes as the way we do that is not "officially" supported. Probably good to investigate that now. I'll have a look on Monday!

seadowg committed:
Don't use broken appcompat version

seadowg commented:
@mmarciniak90 I've fixed the langauge change problem. It was related to a broken Android dependency that got pulled in again (I've added an issue at #3649). 

I'm not sure I can work out a fix for the subscriber id/sim serial changing styles. It seems to be a wierd piece of behaviour in the Android preferences library that occurs only when the preference value is blank (in this case because the phone doesn't have a sim inserted). If you do think it'll case problems for people we could have those fields display "No sim inserted" or "Not set" when there isn't a value to display. An alternative to that would just be hiding the fields when we can't get them.

mmarciniak90 commented:
@seadowg I completely agree! Displaying this information doesn't affect the user experience but allows us to avoid unexpected behavior.
I really like this idea but the final decision belongs to @lognaturel

lognaturel commented:
Using `Not set` wherever a value is unavailable sounds great to me. I think there are a bunch of reasons we might not be able to get the values (permissions, hardware, etc) so I wouldn't want to get too specific. I think it's nice to still display the fields so users know what values could potentially be available.

seadowg commented:
@lognaturel sounds good to me. I'll stick that in this PR

seadowg committed:
Show not set message when device details aren't available

seadowg committed:
Make sure only installID is copyable

seadowg commented:
@lognaturel that's fixed now. I've also made it so only the "Install ID" has the copy on long press behaviour. Initially I was a little unsure how best to deal with the fact that people might end up copying "Not set" somewhere but then I realized we probably want to discourage folks using those other IDs anyway so making "Install ID" the easiest to use seemed like a nice move.

lognaturel commented:
That makes sense. Sorry, just thought of something. I was thinking that "Not set" came from somewhere Android but it doesn't look like it. I'm wondering whether "Not available" would actually make more sense?

seadowg commented:
The standard behaviour (which uses SimpleSummaryProvider) will display "No set" but be a use these preferences are just displayed and not actually stored we're not using that behaviour.

I do like "Not available" a whole lot more though so I'll change it!

seadowg committed:
Improve preference not available string

seadowg commented:
@lognaturel ok that's changed now!

mmarciniak90 commented:
@seadowg I noticed a small issue. I compared PR with the master branch and I noticed that keyboard with alphabet appears instead of a numeric keyboard when I click on `Phone number`

| master| PR|
|---|---|
|<img height="70%" width="70%" src="https://user-images.githubusercontent.com/3907554/75226241-5af1aa00-57ac-11ea-9105-e276379c8844.png">|![Screenshot_2020-02-25-08-49-20-963_org odk collect android](https://user-images.githubusercontent.com/3907554/75226238-5a591380-57ac-11ea-9676-a20bfd7c9b7b.png)|

The second thing that I want to present is a comparison of the PRs appearance. Adding subtext caused the main text is grey out in the Device-defined part. I just wanted to mention this visual change.

|master| PR|
|---|---|
|![Screenshot_20200225-085108](https://user-images.githubusercontent.com/3907554/75227707-fe43be80-57ae-11ea-97f8-e9f0c91f21cf.png)|![Screenshot_20200225-085105](https://user-images.githubusercontent.com/3907554/75227717-026fdc00-57af-11ea-8b3d-4d934347613c.png)|

seadowg commented:
@mmarciniak90 huh interesting on the keyboard - must be a change about how we specificy it's a "number" preference. I'll look at that.

On they grey text, yeah it seem to be the stye that non editable/selectable are disabled in the new preference framework. We might be able to look at changing that in the future but I think we can live with it for this screen.

seadowg committed:
Use numeric keyboard for phone preference

seadowg commented:
@mmarciniak90 fixed! Looks like the new preferences framework had made that less easy to do. Phone number will be using a numeric keyboard now.

mmarciniak90 commented:
**Tested with success**

Verified on Android: 5.1, 6.0, 7.0, 8.1, 9.0, 10.0
Verified on devices with and without a SIM card.

Verified cases:
- Copy Install ID
- Toast
- Reset
- Light/ dark mode
- English, Spanish, RTL language
- Rotating device
- Numeric keyboard for `Phone number`

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
